cudaaligner/src/hirschberg_myers_gpu.cu

```diff
@@ -505,80 +483,159 @@ __device__ void hirschberg_myers_single_char_warp(int8_t*& path, int32_t* path_l
+                ++(data_->cur_end_);
             }
-            __syncwarp(warp_mask);
-            cur_end_ += sizeof(T);
-            return true;
+            success = true;
         }
         else
         {
-            if (threadIdx.x == 0)
+            if (threadIdx.x % warp_size == 0)
             {
                 printf("ERROR: stack full!");
             }
-            return false;
         }
+        __syncwarp();
+        release_mutex();
+        return success;
     }
 
-    __device__ inline void pop()
+    __device__ inline bool pop(T& element)
     {
-        assert(cur_end_ > buffer_begin_);
-        if (cur_end_ - sizeof(T) >= buffer_begin_)
-            cur_end_ -= sizeof(T);
+        lock_mutex();
+        if (threadIdx.x % warp_size == 0)
+        {
+            assert(data_->active_warps_ > 0);
+            --(data_->active_warps_);
+        }
+        release_mutex();
+        bool result = true;
+        while (1)
+        {
+            lock_mutex();
+            __syncwarp();
+            if (data_->buffer_begin_ < data_->cur_end_)
+            {
+                element = *(data_->cur_end_ - 1);
+                __syncwarp();
+                if (threadIdx.x % warp_size == 0)
+                {
+                    --(data_->cur_end_);
+                    ++(data_->active_warps_);
+                }
+                break;
+            }
+            else if (data_->active_warps_ <= 0)
+            {
+                assert(data_->buffer_begin_ == data_->cur_end_);
+                result = false;
+                break;
+            }
+            __syncwarp();
+            release_mutex();
+#if __CUDA_ARCH__ >= 700
+            asm("nanosleep.u32 100000;");
+#endif
+        }
+        __syncwarp();
+        release_mutex();
+        return result;
     }
 
-    __device__ inline T back() const
+private:
+    __device__ inline void release_mutex()
+    {
+        __threadfence_block();
+        if (threadIdx.x % warp_size == 0)
+        {
+            atomicAnd(&(data_->mutex_), 0x0000'0002u);
+        }
+    };
+
+    __device__ inline void lock_mutex()
     {
-        assert(cur_end_ - sizeof(T) >= buffer_begin_);
-        T t;
-        std::memcpy(&t, cur_end_ - sizeof(T), sizeof(T));
-        return t;
+        if (threadIdx.x % warp_size == 0)
+        {
+            while (0 != atomicCAS(&(data_->mutex_), 0, 1))
+            {
+#if __CUDA_ARCH__ >= 700
+                asm("nanosleep.u32 10000;");
+#endif
+            };
+        }
+        __threadfence_block();
     }
 
-    __device__ inline bool empty() const
+    __device__ inline void lock_mutex_high_priority() const
     {
-        return buffer_begin_ == cur_end_;
+        if (threadIdx.x % warp_size == 0)
+        {
+            atomicOr(&(data_->mutex_), 0x0000'0002u); // reserve mutex
+            while (2 != atomicCAS(&(data_->mutex_), 2, 1))
+            {
+#if __CUDA_ARCH__ >= 700
+                asm("nanosleep.u32 1000;");
```

Comment:
how do you decide the sleep time for these while loop? the ones above have much longer sleeps

Reply:
Okay, I think I see now. Since this is the high priority version, you want the loop to check more often that the normal lock, hence the sleep time is smaller.

Reply:
Yes, exactly, I want to be a bit more greedy to get the lock.